merlin/models/tf/models/base.py

```diff
@@ -666,6 +657,18 @@ def __init__(
         ]
         self.schema = sum(input_block_schemas, Schema())
 
+    def _maybe_build(self, inputs):
+        if isinstance(inputs, dict):
+            _ragged_inputs = AsRaggedFeatures()(inputs)
+            feature_shapes = {k: v.shape for k, v in _ragged_inputs.items()}
+            feature_dtypes = {k: v.dtype for k, v in _ragged_inputs.items()}
+
+            for block in self.blocks:
+                for child in block.submodules:
+                    child._feature_shapes = feature_shapes
```

Comment:
📝 this adds metadata to layers at build time, intended to be accessible in compute_output_shape for example.